File: graph_dfs_problems/03-05_trail_one_step5_re.py
# トレイルの通れない頂点 (paizaランク B 相当)
# https://paiza.jp/works/mondai/graph_dfs_problems/graph_dfs__trail_one_step5

INPUT1 = """\
5 1 4
1
3
2
2 5
3
1 3 5
3
2 4 5
2
3 5
4
1 2 3 4
"""
OUTPUT1 = """\
1 2 5 4
"""

INPUT2 = """\
4 1 4
1
2
1
2
3
1 3 4
2
2 4
2
2 3
"""
OUTPUT2 = """\
-1
"""

INPUT3 = """\
6 5 6
1
1
5
2 3 4 5 6
5
1 3 4 5 6
5
1 2 4 5 6
5
1 2 3 5 6
5
1 2 3 4 6
5
1 2 3 4 5
"""
OUTPUT3 = """\
5 4 6 2 4 3 5 2 3 6
"""


def dfs(cv, nodes, edges):
    for nv in ad_list[cv]:
        # 通過済み経路はスキップ
        if edges[cv][nv]:
            continue
        if edges[nv][cv]:
            continue
        # 通れない頂点はスキップ
        if cv in unused_vertices:
            continue

        # nv を通る
        nodes.append(nv)
        edges[cv][nv] = True
        edges[nv][cv] = True

        if nv == t:
            # t に着いたら経路を記録
            global results
            if len(nodes) > len(results[-1]):
                # resuts の頂点数より多いなら更新
                results = [nodes.copy()]
            elif len(nodes) == len(results[-1]):
                results.append(nodes.copy())
        else:
            # 再帰呼び出し
            dfs(nv, nodes, edges)
        # nv を通らない
        nodes.pop()
        edges[cv][nv] = False
        edges[nv][cv] = False


input_str = open(0).read()
input_lines = input_str.splitlines()
# n: 頂点数, s: 起点, t: 終点
n, s, t = map(int, input_lines[0].split())
# 通らない頂点
unused_vertices = list(map(int, input_lines[2].split()))
# 隣接リスト
ad_list = {}
for (i, line) in enumerate(input_lines[3:], 1):
    if i % 2 != 0:
        continue
    ad_list[i // 2] = list(map(int, line.split()))

# s から t への経路
results = [[]]
edges = [[False for w in range(n + 1)] for h in range(n + 1)]
dfs(s, [s], edges)

# s から t に行ける経路を頂点数で昇順ソート
results.sort(key=lambda x: len(x))
# 頂点数が一番多い経路または -1 を出力
if len(results[-1]) > 0:
    print(" ".join(map(str, results[-1])))
else:
    print(-1)

File: graph_dfs_problems/03-05_trail_one_step5_re.rb
# トレイルの通れない頂点 (paizaランク B 相当)
# https://paiza.jp/works/mondai/graph_dfs_problems/graph_dfs__trail_one_step5

INPUT1 = <<~"EOS"
  5 1 4
  1
  3
  2
  2 5
  3
  1 3 5
  3
  2 4 5
  2
  3 5
  4
  1 2 3 4
EOS
OUTPUT1 = <<~"EOS"
  1 2 5 4
EOS

INPUT2 = <<~"EOS"
  4 1 4
  1
  2
  1
  2
  3
  1 3 4
  2
  2 4
  2
  2 3
EOS
OUTPUT2 = <<~"EOS"
  -1
EOS

INPUT3 = <<~"EOS"
  6 5 6
  1
  1
  5
  2 3 4 5 6
  5
  1 3 4 5 6
  5
  1 2 4 5 6
  5
  1 2 3 5 6
  5
  1 2 3 4 6
  5
  1 2 3 4 5
EOS
OUTPUT3 = <<~"EOS"
  5 4 6 2 4 3 5 2 3 6
EOS

def dfs(cv, nodes, edges)
  $ad_list[cv].each do |nv|
    # 通過済み経路はスキップ
    next if edges[cv][nv]
    next if edges[nv][cv]
    # 通れない頂点ならスキップ
    next if unused_vertices.include?(nv)

    # nv を通る
    nodes << nv
    edges[cv][nv] = true
    edges[nv][cv] = true
    # t に着いたら経路を記録
    if nv == $t
      # q を訪問した回数をカウント
      q_count = nodes.count($q)
      if q_count > $results[:q_count]
        # 現在の q_count より多いなら更新
        $results[:q_count] = q_count
        $results[:trails] = [nodes.dup]
      elsif q_count == $results[:q_count] && q_count > 0
        # 現在の q_count と同じなら経路を追加
        $results[:trails] << nodes.dup
      end
    end
    # 再帰呼び出し
    dfs(nv, nodes, edges)
    # nv を通らない
    nodes.pop
    edges[cv][nv] = false
    edges[nv][cv] = false
  end
end

input_str = INPUT1
input_lines = input_str.split("\n")
# n: 頂点数, s: 起点, t: 終点
n, $s, $t = input_lines.shift.split.map(&:to_i)
# 通らない頂点
unused_vertices = input_lines.shift(2).last.split.map(&:to_i)
# 隣接リスト
$ad_list = {}
input_lines.each.with_index(1) do |line, i|
  next if i.odd?
  $ad_list[i / 2] = line.split.map(&:to_i)
end

# s から t への経路
$results = [[]]
edges = Array.new(n + 1) { Array.new(n + 1, false) }
dfs($s, [$s], edges)

# 頂点数が一番多い経路または -1 を出力
results.last.length > 0 ? results.last.join(" ") : -1
